### PR TITLE
[openssl] support fips build feature

### PR DESCRIPTION
**Pull request by lbermes**
I only implemented and tested it for windows as I have no option to do this under linux.
I'm new with changes here but I like to learn and to support this project

- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [X] SHA512s are updated for each updated download
(no checksum changes)
- [X] The "supports" clause reflects platforms that may be fixed by this new version
(I would like to mark the feature as only available for Windows but did not found or understood the syntax how todo this)
- [X] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
(nothing changed)
- [X] Any patches that are no longer applied are deleted from the port's directory.
(nothing changed)
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by lbermes** [comment]
Added "supports" to the feature as I only tested it for windows

**Comment by lbermes** [comment]
> 



> @lbermes please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

File: ports/openssl/portfile.cmake [hunk 1]
```diff
@@ -32,6 +32,12 @@ vcpkg_from_github(
         unix/no-static-libs-for-shared.patch
 )
 
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_CORE_OPTIONS
+FEATURES
+    "fips"    FEATURE_fips
+    )
+
+
```
Comment (Adela0814) [inline ports/openssl/portfile.cmake:40]: Unnecessary, `FEATURE_fips` is not an upstream option.

File: ports/openssl/portfile.cmake [hunk 2]
```diff
@@ -40,6 +46,11 @@ vcpkg_list(SET CONFIGURE_OPTIONS
     no-tests
 )
 
+if("fips" IN_LIST FEATURES)
+    vcpkg_list(APPEND INSTALL_FIPS install_fips)
```
Comment (Adela0814) [inline ports/openssl/portfile.cmake:50]: Do the following targets need to be installed?
```
$ make install_sw
$ make install_ssldirs
$ make install_docs
$ make install_fips     # for `enable-fips` only
```
https://github.com/openssl/openssl/blob/f5b06306b7727f830407c639f8a7bf1698f8aaa5/README-FIPS.md#:~:text=%24%20make%20install_sw%0A%24%20make%20install_ssldirs%0A%24%20make%20install_docs%0A%24%20make%20install_fips%20%20%20%20%20%23%20for%20%60enable%2Dfips%60%20only

**Review by Adela0814 — CHANGES_REQUESTED** [review]

File: ports/openssl/portfile.cmake (hunk 2 above)
Comment (dg0yt) [reply ports/openssl/portfile.cmake:50]: @Adela0814 Why and who do you ask? You know that vcpkg doesn't want to install docs.

File: ports/openssl/vcpkg.json [hunk 3]
```diff
@@ -20,6 +20,10 @@
     }
   ],
   "features": {
+    "fips": {
+      "description": "Enable fips",
+      "supports": "windows"
```
Comment (dg0yt) [inline ports/openssl/vcpkg.json:25]: `supports` is to match upstream's position. We should help you to have all platforms supported.

File: ports/openssl/portfile.cmake (hunk 2 above)
Comment (dg0yt) [reply ports/openssl/portfile.cmake:50]: FWIW `INSTALL_FIPS` should be initialized to an empty value before that `if` block.

**Review by dg0yt — COMMENTED** [review]
This PR shouldn't be merged with a windows-only change, now that windows and !windows finally share most of the openssl setup. But I will need a few more days to look into a general solution.

File: ports/openssl/vcpkg.json (hunk 3 above)
Comment (Adela0814) [reply ports/openssl/vcpkg.json:25]: You can remove `supports` field, I have tested `fips `on `linux` and `osx`.

**Review by Adela0814 — CHANGES_REQUESTED** [review]

**Review by Adela0814 — DISMISSED** [review]

**Review by dg0yt — CHANGES_REQUESTED** [review]
I tested the PR.
AFAIU fips support is only available with dynamic linkage. (There is no build error, but the feature will simply not be installed.) The feature should be declared with:
~~~json
"supports": "!static"
~~~

**Review by dg0yt — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks!

**Comment by lbermes** [comment]
Thank you for your support

**Comment by BillyONeal** [comment]
> Should I simply create a new pull request with the fix?

Yes please

**Comment by dg0yt** [comment]
While we are at fips: Should it become a default feature? (Is it a default feature upstream?)

**Comment by BillyONeal** [comment]
> While we are at fips: Should it become a default feature? (Is it a default feature upstream?)

I don't think so; normally FIPS only certifies distribution in binary forms, not in source forms. Moreover, I don't know about OpenSSL specifically, but elsewhere I've seen a FIPS mode that usually implies turning things off (that have not been certified) rather than on.

**Comment by guenterbe** [comment]
After getting the latest updates from github and rebuilding openssl, I can't find any fips.dll on Windows or fips.so on Linux.
Isn't a "vcpkg install openssl" sufficient? Are there some undocumented (or not found by me) command line switches?

**Comment by lbermes** [comment]
Hi, please use the feature fips.
I added this first only for windows but the see the comment of [Adela0814] 
It was requested that I remove this and so it now is a feature for all platforms.

**Comment by lbermes** [comment]
"vcpkg install openssl[fips]" or with the tools "vcpkg install openssl[tools,fips]"

**Comment by lbermes** [comment]
But I wonder how this influenced the linux build. I did not adjusted the linux port file.

**Comment by guenterbe** [comment]
@lbermes I have never needed the [features] in the past and was therefore not aware of them (must have missed them in the docs). Thanks very much!

**Pull request merged**